File: src/lib/strategy-evolver/stability.ts
import type { MarketDataInput, StrategyParameterSet, StabilityReport } from './types';
import { extractReturnsForSymbols, scoreParameterSet } from './walk-forward';

function deepCloneParams(p: StrategyParameterSet): StrategyParameterSet {
  return {
    ...p,
    safeAllocation: { ...p.safeAllocation },
    ambitionAllocation: { ...p.ambitionAllocation },
  };
}

function perturbWeights(
  weights: { [symbol: string]: number },
  symbol: string,
  delta: number,
): { [symbol: string]: number } {
  const result = { ...weights };
  const current = result[symbol] ?? 0;
  const adjusted = Math.max(0, Math.min(1, current + delta));
  result[symbol] = adjusted;

  const otherSymbols = Object.keys(result).filter(k => k !== symbol);
  const otherTotal = otherSymbols.reduce((s, k) => s + (result[k] ?? 0), 0);
  const diff = adjusted - current;

  if (otherTotal > 0 && otherSymbols.length > 0) {
    for (const k of otherSymbols) {
      result[k] = (result[k] ?? 0) - diff * ((result[k] ?? 0) / otherTotal);
    }
  }

  const sum = Object.values(result).reduce((s, v) => s + v, 0);
  if (sum > 0) {
    for (const k of Object.keys(result)) {
      result[k] = (result[k] ?? 0) / sum;
    }
  }

  return result;
}

export function checkStability(
  data: MarketDataInput,
  symbols: string[],
  optimalParams: StrategyParameterSet,
  radius: number = 0.05,
  threshold: number = 0.1,
  riskFreeRate: number = 0,
  testRatio: number = 0.3,
): StabilityReport {
  const allReturns = extractReturnsForSymbols(data, symbols);
  if (allReturns.length === 0) {
    return { gradient: 1, threshold, isStable: false, neighborhoodSharpeRatios: [] };
  }

  const totalObs = allReturns[0].length;
  const testStart = Math.floor(totalObs * (1 - testRatio));
  const testEnd = totalObs - 1;

  const baseScore = scoreParameterSet(
    symbols, allReturns, testStart, testEnd, optimalParams, riskFreeRate,
  );

  const neighborhoodSharpeRatios: number[] = [baseScore];
  const gradients: number[] = [];

  const scalarParams: Array<{ key: keyof StrategyParameterSet; get: (p: StrategyParameterSet) => number; set: (p: StrategyParameterSet, v: number) => void; isInt?: boolean }> = [
    {
      key: 'triggerLine',
      get: (p) => p.triggerLine,
      set: (p, v) => { p.triggerLine = Math.round(v); },
      isInt: true,
    },
    {
      key: 'safeRatio',
      get: (p) => p.safeRatio,
      set: (p, v) => { p.safeRatio = Math.max(0, Math.min(1, v)); },
    },
    {
      key: 'ambitionRatio',
      get: (p) => p.ambitionRatio,
      set: (p, v) => { p.ambitionRatio = Math.max(0, Math.min(1, v)); },
    },
    {
      key: 'bsmThreshold',
      get: (p) => p.bsmThreshold,
      set: (p, v) => { p.bsmThreshold = Math.max(0, v); },
    },
    {
      key: 'maShortWindow',
      get: (p) => p.maShortWindow,
      set: (p, v) => { p.maShortWindow = Math.max(1, Math.round(v)); },
      isInt: true,
    },
    {
      key: 'maLongWindow',
      get: (p) => p.maLongWindow,
      set: (p, v) => { p.maLongWindow = Math.max(1, Math.round(v)); },
      isInt: true,
    },
  ];

  for (const param of scalarParams) {
    const baseVal = param.get(optimalParams);
    if (baseVal === 0) continue;

    const perturbAmount = Math.max(param.isInt ? 1 : radius, Math.abs(baseVal * radius));
    if (perturbAmount === 0) continue;

    const pUp = deepCloneParams(optimalParams);
    param.set(pUp, baseVal + perturbAmount);
    const scoreUp = scoreParameterSet(
      symbols, allReturns, testStart, testEnd, pUp, riskFreeRate,
    );
    neighborhoodSharpeRatios.push(scoreUp);
    gradients.push(Math.abs(scoreUp - baseScore) / perturbAmount);

    const pDown = deepCloneParams(optimalParams);
    param.set(pDown, Math.max(0, baseVal - perturbAmount));
    const scoreDown = scoreParameterSet(
      symbols, allReturns, testStart, testEnd, pDown, riskFreeRate,
    );
    neighborhoodSharpeRatios.push(scoreDown);
    gradients.push(Math.abs(scoreDown - baseScore) / perturbAmount);
  }

  const safeSymbols = Object.keys(optimalParams.safeAllocation);
  for (const sym of safeSymbols) {
    const baseW = optimalParams.safeAllocation[sym] ?? 0;
    const delta = Math.max(radius, baseW > 0 ? baseW * radius : radius);

    const pUp = deepCloneParams(optimalParams);
    pUp.safeAllocation = perturbWeights(pUp.safeAllocation, sym, delta);
    const scoreUp = scoreParameterSet(
      symbols, allReturns, testStart, testEnd, pUp, riskFreeRate,
    );
    neighborhoodSharpeRatios.push(scoreUp);
    gradients.push(Math.abs(scoreUp - baseScore) / delta);

    const pDown = deepCloneParams(optimalParams);
    pDown.safeAllocation = perturbWeights(pDown.safeAllocation, sym, -delta);
    const scoreDown = scoreParameterSet(
      symbols, allReturns, testStart, testEnd, pDown, riskFreeRate,
    );
    neighborhoodSharpeRatios.push(scoreDown);
    gradients.push(Math.abs(scoreDown - baseScore) / delta);
  }

  const ambitionSymbols = Object.keys(optimalParams.ambitionAllocation);
  for (const sym of ambitionSymbols) {
    const baseW = optimalParams.ambitionAllocation[sym] ?? 0;
    const delta = Math.max(radius, baseW > 0 ? baseW * radius : radius);

    const pUp = deepCloneParams(optimalParams);
    pUp.ambitionAllocation = perturbWeights(pUp.ambitionAllocation, sym, delta);
    const scoreUp = scoreParameterSet(
      symbols, allReturns, testStart, testEnd, pUp, riskFreeRate,
    );
    neighborhoodSharpeRatios.push(scoreUp);
    gradients.push(Math.abs(scoreUp - baseScore) / delta);

    const pDown = deepCloneParams(optimalParams);
    pDown.ambitionAllocation = perturbWeights(pDown.ambitionAllocation, sym, -delta);
    const scoreDown = scoreParameterSet(
      symbols, allReturns, testStart, testEnd, pDown, riskFreeRate,
    );
    neighborhoodSharpeRatios.push(scoreDown);
    gradients.push(Math.abs(scoreDown - baseScore) / delta);
  }

  const avgGradient = gradients.length > 0
    ? gradients.reduce((s, g) => s + g, 0) / gradients.length
    : 0;

  return {
    gradient: avgGradient,
    threshold,
    isStable: avgGradient < threshold,
    neighborhoodSharpeRatios,
  };
}

File: scripts/local-evolver/stability.py
"""Parameter stability check via per-parameter neighborhood gradient.

Perturbs each parameter individually (up/down) and measures:
  gradient_i = |score_up - score_base| / |param_up - param_base|

Averages all gradients and compares against threshold.
This prevents sensitivity in individual parameters from being masked
by cancellation in simultaneous random perturbations.
"""

import copy

import numpy as np

from dsr import compute_sharpe_ratio
from models import MarketDataInput, StabilityReport, StrategyParameterSet
from walk_forward import compute_portfolio_returns_for_params, extract_returns_for_symbols


def _perturb_weights(
    weights: dict[str, float],
    symbol: str,
    delta: float,
) -> dict[str, float]:
    result = dict(weights)
    current = result.get(symbol, 0.0)
    adjusted = max(0.0, min(1.0, current + delta))
    result[symbol] = adjusted

    other_keys = [k for k in result if k != symbol]
    other_total = sum(result[k] for k in other_keys)
    diff = adjusted - current

    if other_total > 0 and len(other_keys) > 0:
        for k in other_keys:
            result[k] = result[k] - diff * (result[k] / other_total)

    total = sum(result.values())
    if total > 0:
        for k in result:
            result[k] /= total

    return result


def check_stability(
    data: MarketDataInput,
    symbols: list[str],
    params: StrategyParameterSet,
    neighborhood_radius: float = 0.05,
    gradient_threshold: float = 0.1,
    risk_free_rate: float = 0.0,
    test_ratio: float = 0.3,
) -> StabilityReport:
    all_rets = extract_returns_for_symbols(data, symbols)
    if not all_rets:
        return StabilityReport(gradient=1.0, threshold=gradient_threshold, is_stable=False)

    total_obs = len(all_rets[0])
    test_start = int(total_obs * (1 - test_ratio))
    test_end = total_obs - 1

    base_rets = compute_portfolio_returns_for_params(
        symbols, all_rets, test_start, test_end, params,
    )
    base_sharpe = compute_sharpe_ratio(base_rets, risk_free_rate) if len(base_rets) >= 5 else -1.0

    neighborhood_sharpes: list[float] = [base_sharpe]
    gradients: list[float] = []

    # Scalar parameters
    scalar_params: list[tuple[str, float | int]] = [
        ("trigger_line", params.trigger_line),
        ("safe_ratio", params.safe_ratio),
        ("ambition_ratio", params.ambition_ratio),
        ("bsm_threshold", params.bsm_threshold),
        ("ma_short_window", params.ma_short_window),
        ("ma_long_window", params.ma_long_window),
    ]

    for _name, base_val in scalar_params:
        if base_val == 0:
            continue
        if isinstance(base_val, int):
            perturb = max(1, int(abs(base_val) * neighborhood_radius))
        else:
            perturb = max(neighborhood_radius, abs(base_val) * neighborhood_radius)
        if perturb == 0:
            continue

        p_up = copy.deepcopy(params)
        new_val_up = base_val + perturb
        if isinstance(base_val, int):
            setattr(p_up, _name, int(round(new_val_up)))
        else:
            setattr(p_up, _name, new_val_up)
        pr_up = compute_portfolio_returns_for_params(
            symbols, all_rets, test_start, test_end, p_up,
        )
        score_up = compute_sharpe_ratio(pr_up, risk_free_rate) if len(pr_up) >= 5 else -1.0
        neighborhood_sharpes.append(score_up)
        gradients.append(abs(score_up - base_sharpe) / perturb)

        p_down = copy.deepcopy(params)
        new_val_down = max(0, base_val - perturb)
        if isinstance(base_val, int):
            setattr(p_down, _name, int(round(new_val_down)))
        else:
            setattr(p_down, _name, new_val_down)
        pr_down = compute_portfolio_returns_for_params(
            symbols, all_rets, test_start, test_end, p_down,
        )
        score_down = compute_sharpe_ratio(pr_down, risk_free_rate) if len(pr_down) >= 5 else -1.0
        neighborhood_sharpes.append(score_down)
        gradients.append(abs(score_down - base_sharpe) / perturb)

    # Allocation weights
    for sym in params.safe_allocation:
        base_w = params.safe_allocation[sym]
        delta = max(neighborhood_radius, abs(base_w) * neighborhood_radius)

        p_up = copy.deepcopy(params)
        p_up.safe_allocation = _perturb_weights(p_up.safe_allocation, sym, delta)
        pr_up = compute_portfolio_returns_for_params(
            symbols, all_rets, test_start, test_end, p_up,
        )
        score_up = compute_sharpe_ratio(pr_up, risk_free_rate) if len(pr_up) >= 5 else -1.0
        neighborhood_sharpes.append(score_up)
        gradients.append(abs(score_up - base_sharpe) / delta)

        p_down = copy.deepcopy(params)
        p_down.safe_allocation = _perturb_weights(p_down.safe_allocation, sym, -delta)
        pr_down = compute_portfolio_returns_for_params(
            symbols, all_rets, test_start, test_end, p_down,
        )
        score_down = compute_sharpe_ratio(pr_down, risk_free_rate) if len(pr_down) >= 5 else -1.0
        neighborhood_sharpes.append(score_down)
        gradients.append(abs(score_down - base_sharpe) / delta)

    for sym in params.ambition_allocation:
        base_w = params.ambition_allocation[sym]
        delta = max(neighborhood_radius, abs(base_w) * neighborhood_radius)

        p_up = copy.deepcopy(params)
        p_up.ambition_allocation = _perturb_weights(p_up.ambition_allocation, sym, delta)
        pr_up = compute_portfolio_returns_for_params(
            symbols, all_rets, test_start, test_end, p_up,
        )
        score_up = compute_sharpe_ratio(pr_up, risk_free_rate) if len(pr_up) >= 5 else -1.0
        neighborhood_sharpes.append(score_up)
        gradients.append(abs(score_up - base_sharpe) / delta)

        p_down = copy.deepcopy(params)
        p_down.ambition_allocation = _perturb_weights(p_down.ambition_allocation, sym, -delta)
        pr_down = compute_portfolio_returns_for_params(
            symbols, all_rets, test_start, test_end, p_down,
        )
        score_down = compute_sharpe_ratio(pr_down, risk_free_rate) if len(pr_down) >= 5 else -1.0
        neighborhood_sharpes.append(score_down)
        gradients.append(abs(score_down - base_sharpe) / delta)

    avg_gradient = np.mean(gradients) if gradients else 1.0

    return StabilityReport(
        gradient=float(avg_gradient),
        threshold=gradient_threshold,
        is_stable=bool(avg_gradient < gradient_threshold),
        neighborhood_sharpe_ratios=neighborhood_sharpes,
    )
